omniqueue/src/lib.rs

```diff
@@ -1,3 +1,101 @@
+//! # Omniqueue
+//!
+//! Omniqueue is an abstraction layer over various queue backends. As of present it supports:
+//!
+//!   * In-memory queues
+//!
+//!   * RabbitMQ
+//!
+//!   * Redis streams
+//!
+//!   * SQS
+//!
+//! Omniqueue provides a high-level interface for sending and receiving
+//!
+//!   * Raw byte arrays in the way most compatible with the queue backend
+//!
+//!   * JSON encoded byte arrays for types that implement [`serde::Deserialize`] and
+//!     [`serde::Serialize`]
+//!
+//!   * Arbitrary types for which an encoder and/or decoder has been defined
+//!
+//! ## How to Use Omniqueue
+//!
+//!   1. Start by adding `omniqueue` to your `Cargo.toml`. If you don't need support for all of the
+//!      backends, disable the default features and enable each required backend individually.
+//!
+//!   2. Create the configuration for your queue backend. Depending on which backend you use, this
+//!      could be [`backends::rabbitmq::RabbitMqConfig`], [`backends::redis::RedisConfig`],
+//!      [`backends::sqs::SqsConfig`], or a simple [`usize`] for memory queues.
+//!
+//!   3. Create the producer and/or consumer:
+//!
+//!      ```compile_fail
+//!      let cfg = SqsConfig {
+//!          queue_dsn: "http://localhost:9234/queue/queue_name".to_owned(),
+//!          override_endpoint: true,
+//!      };
+//!      
+//!      // Either both producer and consumer
+//!      let (p, mut c) = SqsQueueBackend::builder(cfg.clone()).build_pair().await?;
+//!
+//!      // Or one half
+//!      let p = SqsQueueBackend::builder(cfg.clone()).build_producer().await?;
+//!      let mut c = SqsQueueBackend::builder(cfg).build_consumer().await?;
+//!
+//!      (p, c)
+//!      ```
+//!
+//!   4. Send and receive information:
+//!
+//!      ```compile_fail
+//!      p.send_serde_json(&ExampleType::default()).await?;
+//!
+//!      let delivery = c.receive().await?;
+//!      let payload = delivery.payload_serde_json::<ExampleType>().await?;
+//!      delivery.ack().await?;
+//!      ```
+//! ## `DynProducer`s and `DynConsumer`s
+//!
+//! Dynamic-dispatch can be used easily for when you're not sure which backend is to be used at
+//! compile-time.
```

Comment:
This is interesting. I thought dynamic dyspatch was for when you don't know the type at _runtime_. Boxed, opaque, whatever.

Reply:
I guess it depends on how you define "know". With `std::any::Any` and downcasting, it could be argued that trait objects are usually "known" at runtime (although tracking that is probably optimized away if you're not using it).

I wrote "at compile-time" because of the cases like how I'm trying to use the crate in https://github.com/svix/svix-webhooks i.e. where the type is configured so it's "concrete type" is only ever known as it's being constructed based on a branch entered dependent on a config file. But the type _is_ known at runtime based on the enum in the config?

I'm not sure which verbiage is more correct to be honest.